tokio-util/src/task/join_map.rs

```diff
@@ -626,6 +626,19 @@ where
         }
     }
 
+    /// Returns an iterator visiting all keys in this `JoinMap` in arbitrary order.
+    /// If a task has completed, but its output hasn't yet been consumed by a
+    /// call to [`join_next`], this method will still return its key.
+    ///
+    /// [`join_next`]: fn@Self::join_next
+    pub fn keys<Q: ?Sized>(&self) -> impl Iterator<Item = &K>
```

Comment:
it's definitely much simpler to return an `impl Iterator` here. However, I will note that the `hash_map::Keys` iterator that this wraps also implements some other traits: in addition to being an `Iterator`, [it also implements `Clone`, `ExactSizeIterator`, and `FusedIterator`](https://doc.rust-lang.org/stable/std/collections/hash_map/struct.Keys.html#trait-implementations). It might be better if we returned an iterator that also implements those traits. 

In particular, `ExactSizeIterator` is used by the standard library to implement optimizations when `collect()`ing an iterator into particular collections, like `Vec`s and `HashMap`s, since the size of the iterator can be reserved up front by `with_capacity`, instead of allocating multiple times while collecting.

Therefore, we may want to change this to return a named `Keys` iterator, like this:

```rust
use std::iterator::{ExactSizeIterator, FusedIterator};

pub struct Keys<'map, K, S = RandomState> {
    inner: hashbrown::hash_map::Keys<'map, Key<K>, AbortHandle, S>>,
}

impl<'map, K, S> Iterator for Keys<'map, K, S> {
    type Item = &'map K;

    fn next(&mut self) -> Option<Self::Item> {
         self.inner.next.map(|key| &key.key)
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        self.inner.size_hint()    
    }
}

impl<'map, K, S> ExactSizeIterator for Keys<'map, K, S> {
    fn len(&self) -> usize {
         self.inner.len()
    }

    fn is_empty(&self) -> bool {
         self.inner.is_empty()
    }
}

impl<'map, K, S> FusedIterator for Keys<'map, K, S> {}
```

Or, alternatively, I think we could write
```suggestion
    pub fn keys<Q: ?Sized>(&self) -> impl Iterator<Item = &K> + ExactSizeIterator + FusedIterator
```

but I think the named type might be a nicer API surface for users?